src/hotspot/share/cds/dynamicArchive.hpp

```diff
@@ -38,12 +38,12 @@
 #if INCLUDE_CDS
 
 class DynamicArchiveHeader : public FileMapHeader {
-  friend class CDSOffsets;
 private:
   int _base_header_crc;
-  int _base_region_crc[MetaspaceShared::n_regions];
 
 public:
+  int _base_region_crc[MetaspaceShared::n_regions];
```

Comment:
This looks like it should be private with either a public accessor or else a friend declaration for the client code.

Reply:
Let me try but I remember it does not work with constexpr initialization. This is why I move the data field to public not using the public accessor.

Reply:
So we are trading off use of constexpr versus broken encapsulation? :( I can't say I understand the details here.

Reply:
For constexpr array --- the initialization requires every item in the array must be initialized with const expression. I have rolled back the access change. By putting the global arrays offsets and constants inside CDSConstants, and making CDSConstants a friend class of the accessed class, we can avoid such changes. Thanks!

Reply:
`_base_region_crc` should be private and `CDSConstants` should be a friend class of `DynamicArchiveHeader`